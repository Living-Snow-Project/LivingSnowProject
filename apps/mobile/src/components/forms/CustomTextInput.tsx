import React, { forwardRef /* useRef */ } from "react";
import { FormControl, Input, WarningOutlineIcon } from "native-base";
/* import {
  NativeSyntheticEvent,
  TextInputContentSizeChangeEventData,
} from "react-native";
import RCTDeviceEventEmitter from "react-native/Libraries/EventEmitter/RCTDeviceEventEmitter"; */

type CustomTextInputProps = {
  label: string;
  placeholder: string;
  value?: string | undefined;
  validation?: string;
  maxLength?: number;
  isDisabled?: boolean;
  isInvalid?: boolean;
  isRequired?: boolean;
  onChangeText?: (text: string) => void;
  onSubmitEditing?: () => void;
};

export const CustomTextInput = forwardRef<typeof Input, CustomTextInputProps>(
  (
    {
      label,
      placeholder,
      value,
      validation,
      maxLength = 255,
      isDisabled = false,
      isInvalid = false,
      isRequired = false,
      onChangeText = () => {},
      onSubmitEditing = () => {},
    }: CustomTextInputProps,
    ref
  ) => (
    /* const textInputHeight = useRef(0);

    // onContentSizeChange is called frequently for multiline TextInput, we only want to emit 'keyboardDidShow' event when height actually changes
    const handleMultilineTextInputOnContentSizeChange = (
      height: number,
      event: NativeSyntheticEvent<TextInputContentSizeChangeEventData>
    ) => {
      if (height !== 0 && height !== event.nativeEvent.contentSize.height) {
        // 'keyboardDidShow' expects the height of the keyboard (which we could capture in a new event listener in this component)
        // since we only have 1 'keyboardDidShow' listener we changed its logic to respond to this input
        // this is potentially bad if we set up additional 'keyboardDidShow' listeners in the app
        RCTDeviceEventEmitter.emit("keyboardDidShow", {});
      }

      return event.nativeEvent.contentSize.height;
    }; */

    <FormControl isRequired={isRequired} isInvalid={isInvalid}>
      <FormControl.Label>{label}</FormControl.Label>
      <Input
        size="lg"
        multiline
        blurOnSubmit
        // @ts-ignore ref typing isn't quite right in native-base yet
        ref={ref}
        value={value}
        placeholder={placeholder}
        isDisabled={isDisabled}
        onChangeText={onChangeText}
        onSubmitEditing={onSubmitEditing}
        /* onContentSizeChange={(event) => {
          textInputHeight.current =
            handleMultilineTextInputOnContentSizeChange(
              textInputHeight.current,
              event
            );
        }} */
        maxLength={maxLength}
        returnKeyType="done"
        variant="outline"
      />
      <FormControl.ErrorMessage leftIcon={<WarningOutlineIcon size="sm" />}>
        {validation}
      </FormControl.ErrorMessage>
    </FormControl>
  )
);

// this is needed because of forwardRef
CustomTextInput.displayName = `CustomTextInput`;
