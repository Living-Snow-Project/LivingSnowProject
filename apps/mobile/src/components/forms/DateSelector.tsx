import React, { useState } from "react";
import { View } from "react-native";
import { FormControl, Input, Pressable, useColorModeValue } from "native-base";
import { Calendar } from "react-native-calendars";
import { Labels } from "../../constants";

type DateSelectorProps = {
  date: string;
  maxDate: string;
  setDate: (date: string) => void;
};

const dark = "#44403c"; // NativeBase light.700
const light = "#fafaf9"; // NativeBase light.50

export function DateSelector({ date, maxDate, setDate }: DateSelectorProps) {
  const [calendarVisible, setCalendarVisible] = useState(false);
  const bgColor = useColorModeValue(light, dark);
  const dayColor = useColorModeValue(dark, light);
  const selectedDayColor = bgColor;
  const monthColor = dayColor;
  const disabledColor = useColorModeValue(`${dark}55`, `${light}55`);

  const renderCalendar = () => {
    if (calendarVisible) {
      return (
        <Calendar
          testID="calendar"
          current={date}
          onDayPress={(newDate) => {
            setCalendarVisible(false);
            setDate(newDate.dateString);
          }}
          maxDate={maxDate}
          markedDates={{ [date]: { selected: true } }}
          theme={{
            calendarBackground: bgColor,
            dayTextColor: dayColor,
            selectedDayTextColor: selectedDayColor,
            monthTextColor: monthColor,
            textDisabledColor: disabledColor,
          }}
        />
      );
    }

    return (
      <View pointerEvents="none">
        <Input value={date} size="lg" />
      </View>
    );
  };

  return (
    <FormControl isRequired>
      <FormControl.Label>{Labels.Date}</FormControl.Label>
      <Pressable
        testID="calendar-pressable"
        onPress={() => setCalendarVisible(true)}
      >
        {renderCalendar()}
      </Pressable>
    </FormControl>
  );
}
