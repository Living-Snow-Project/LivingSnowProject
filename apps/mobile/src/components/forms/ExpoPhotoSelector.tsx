import React from "react";
import { FormControl, Pressable } from "native-base";
import * as ImagePicker from "expo-image-picker";
import * as MediaLibrary from "expo-media-library";
import { manipulateAsync, ActionResize } from "expo-image-manipulator";
import { SelectedPhoto } from "../../../types";
import { ThemedBox } from "../layout";
import { AddPhotosIcon, PhotosLayout } from "../media";
import { PhotoManager } from "../../lib/PhotoManager";
import { Labels, TestIds } from "../../constants";
import { useToast } from "../../hooks";
import { ToastAlert } from "../feedback";

type ExpoPhotoSelectorProps = {
  recordId: string;
  photos: SelectedPhoto[];
  setSelectedPhotos: (selectedPhotos: SelectedPhoto[]) => void;
  setStatus: (status: "Idle" | "Loading") => void;
};

export function ExpoPhotoSelector({
  recordId,
  photos,
  setSelectedPhotos,
  setStatus,
}: ExpoPhotoSelectorProps) {
  const toast = useToast();

  const handleOnPress = async () => {
    setStatus("Loading");

    try {
      const permission = await MediaLibrary.getPermissionsAsync();

      if (!permission.granted) {
        const granted = await MediaLibrary.requestPermissionsAsync();

        if (!granted.granted) {
          return;
        }
      }

      const result = await ImagePicker.launchImageLibraryAsync({
        mediaTypes: ImagePicker.MediaTypeOptions.Images,
        allowsMultipleSelection: true,
        selectionLimit: 4,
        quality: 0.7,
      });

      // TODO: should "cancel" unselect all? there is no way to go back to 0 selected after first selection
      if (!result.canceled) {
        const assets: MediaLibrary.Asset[] = [];

        await result.assets.reduce(async (promise, current) => {
          await promise;

          if (current.assetId) {
            // returned ImagePickerAsset is missing various fields in Asset that are "needed" (for TypeScript)
            const asset = await MediaLibrary.getAssetInfoAsync(current.assetId);

            // resize to 1024 along major axis
            const dims: ActionResize = { resize: {} };

            if (asset.height > asset.width) {
              dims.resize.height = 1024;
            } else {
              dims.resize.width = 1024;
            }

            const resizedImage = await manipulateAsync(current.uri, [dims]);

            assets.push({
              ...asset,
              height: resizedImage.height,
              uri: resizedImage.uri,
              width: resizedImage.width,
            });
          }

          return Promise.resolve();
        }, Promise.resolve());

        PhotoManager.addSelected(recordId, assets);
        setSelectedPhotos(assets);
      }
    } catch (error) {
      toast.show(
        <ToastAlert
          status="info"
          title="Upload photo error"
          message="We ran into an error preparing photos for upload."
        />,
      );
    } finally {
      setStatus("Idle");
    }
  };

  const renderPhotos = () => {
    if (photos.length) {
      return (
        <ThemedBox mx={-2}>
          <PhotosLayout photos={photos} />
        </ThemedBox>
      );
    }

    return (
      <ThemedBox height="52" width="52">
        <AddPhotosIcon />
      </ThemedBox>
    );
  };

  return (
    <FormControl>
      <FormControl.Label>{Labels.RecordScreen.Photos}</FormControl.Label>
      <Pressable testID={TestIds.Selectors.Photos} onPress={handleOnPress}>
        {renderPhotos()}
      </Pressable>
    </FormControl>
  );
}
