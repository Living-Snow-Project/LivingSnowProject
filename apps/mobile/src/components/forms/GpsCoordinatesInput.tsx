import React, { useEffect, useRef, useState } from "react";
import {
  Alert,
  Pressable,
  ReturnKeyTypeOptions,
  Text,
  TextInput,
  View,
} from "react-native";
import {
  Accuracy,
  requestForegroundPermissionsAsync,
  watchPositionAsync,
} from "expo-location";
import Logger from "@livingsnow/logger";
import { formInputStyles } from "../../styles/FormInput";
import { getAppSettings } from "../../../AppSettings";
import { TestIds } from "../../constants/TestIds";
import { Placeholders } from "../../constants/Strings";

type GpsCoordinates = {
  latitude: number;
  longitude: number;
};

type GpsCoordinatesInputProps = {
  setGpsCoordinates: (latitude: number, longitude: number) => void;
  onSubmitEditing: () => void;
  coordinates?: GpsCoordinates;
};

export function GpsCoordinatesInput({
  setGpsCoordinates,
  onSubmitEditing,
  coordinates = { latitude: 0, longitude: 0 },
}: GpsCoordinatesInputProps) {
  const watchPosition = useRef<null | { remove(): void }>(null);
  const gpsCoordinatesRef = useRef<TextInput>(null);
  const [placeholderText, setPlaceholderText] = useState(
    Placeholders.GPS.AcquiringLocation
  );
  const useGps = coordinates.latitude === 0 && coordinates.longitude === 0;
  const [gpsCoordinateString, setGpsCoordinateString] = useState<string>(
    useGps ? `` : `${coordinates.latitude}, ${coordinates.longitude}`
  );
  const [manualGpsCoordinates, setManualGpsCoordinates] = useState(!useGps);
  const showGpsWarning = useGps ? getAppSettings().showGpsWarning : false;
  const clipCoordinate = (coordinate: number) =>
    JSON.stringify(coordinate.toFixed(6)).replace('"', "").replace('"', "");

  // accepts inputs from both location subscription and typed user input
  const updateGpsCoordinateString = (value: string) => {
    setGpsCoordinateString(value);

    // some users are adding parenthesis when manually entering coordinates
    // we don't want them in the data set so quietly remove
    const coords = value.replace(`(`, ``).replace(`)`, ``).split(",");

    let latitude;
    let longitude;

    if (coords[0]) {
      latitude = Number(coords[0].trim());
    }

    if (coords[1]) {
      longitude = Number(coords[1].trim());
    }

    setGpsCoordinates(latitude, longitude);
  };

  const stopGps = () => {
    if (watchPosition.current) {
      watchPosition.current?.remove();
      watchPosition.current = null;
    }
  };

  // location subscription
  useEffect(() => {
    if (!useGps) return () => {};

    (async () => {
      const { status } = await requestForegroundPermissionsAsync();
      if (status !== "granted") {
        setManualGpsCoordinates(true);
        setPlaceholderText(Placeholders.GPS.NoPermissions);
        Logger.Warn(`Permission to access foreground location was ${status}.`);
        return;
      }

      try {
        watchPosition.current = await watchPositionAsync(
          {
            accuracy: Accuracy.High,
            timeInterval: 5000,
          },
          ({ coords }) =>
            updateGpsCoordinateString(
              `${clipCoordinate(coords.latitude)}, ${clipCoordinate(
                coords.longitude
              )}`
            )
        );
      } catch (error) {
        setManualGpsCoordinates(true);
        setPlaceholderText(Placeholders.GPS.NoLocation);
        Logger.Error(`Location.watchPositionAsync() failed: ${error}`);
      }
    })();

    return stopGps;
  }, []);

  useEffect(() => {
    if (useGps) {
      gpsCoordinatesRef?.current?.focus();
    }
  }, [manualGpsCoordinates]);

  const enableManualGpsCoordinates = () => {
    stopGps();
    updateGpsCoordinateString("");
    setManualGpsCoordinates(true);
    setPlaceholderText(Placeholders.GPS.EnterCoordinates);
  };

  const confirmManualGpsCoordinates = () => {
    if (showGpsWarning) {
      Alert.alert(
        "Enter GPS coordinates manually?",
        "This message can be disabled in Settings.",
        [
          {
            text: "Yes",
            onPress: () => enableManualGpsCoordinates(),
          },
          {
            text: "No",
            style: "cancel",
          },
        ]
      );
    } else {
      enableManualGpsCoordinates();
    }
  };

  const gpsFieldProps = {
    style: formInputStyles.optionInputText,
    editable: manualGpsCoordinates,
    maxLength: 30,
    placeholder: placeholderText,
    defaultValue: gpsCoordinateString,
    returnKeyType: "done" as ReturnKeyTypeOptions,
  };

  return (
    <>
      <Text style={formInputStyles.optionStaticText}>
        GPS Coordinates (latitude, longitude)
      </Text>
      {!manualGpsCoordinates && (
        /* user confirms they want to enter GPS coordinates manually */
        <Pressable
          onPress={() => confirmManualGpsCoordinates()}
          testID={TestIds.GPS.gpsManualPressableTestId}
        >
          <View pointerEvents="none">
            <TextInput {...gpsFieldProps} />
          </View>
        </Pressable>
      )}
      {manualGpsCoordinates && (
        <TextInput
          {...gpsFieldProps}
          testID={TestIds.GPS.gpsManualInputTestId}
          onChangeText={(value) => updateGpsCoordinateString(value)}
          onSubmitEditing={() => onSubmitEditing()}
          ref={gpsCoordinatesRef}
        />
      )}
    </>
  );
}
