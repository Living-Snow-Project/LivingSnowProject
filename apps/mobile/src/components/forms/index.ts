export * from "./CustomTextInput";
export * from "./DateSelector";
export * from "./ExpoPhotoSelector";
export * from "./GpsCoordinatesInput";
export * from "./PhotoSelector";
export * from "./TextArea";
export * from "./TypeSelector";
export * from "./UserIdentityInput";
