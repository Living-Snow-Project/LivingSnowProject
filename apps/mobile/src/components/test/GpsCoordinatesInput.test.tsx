import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import * as Location from "expo-location";
import { GpsCoordinatesInput } from "../forms/GpsCoordinatesInput";
import { Placeholders } from "../../constants";
import { NativeBaseProviderForTesting } from "../../../jesttest.setup";

describe("GpsCoordinatesInput tests", () => {
  const onSubmitEditingMock = jest.fn();
  const setCoordinatesMock = jest.fn();

  const requestForegroundPermissionsAsyncMock = (result: string) =>
    jest
      .spyOn(Location, "requestForegroundPermissionsAsync")
      .mockImplementation(() =>
        Promise.resolve({
          status: result,
        } as Location.LocationPermissionResponse)
      );

  const requestForegroundPermissionsAsyncSuccessMock = () =>
    requestForegroundPermissionsAsyncMock("granted");

  const requestForegroundPermissionsAsyncFailureMock = () =>
    requestForegroundPermissionsAsyncMock("denied");

  const watchPositionAsyncSuccessMock = () =>
    jest
      .spyOn(Location, "watchPositionAsync")
      .mockImplementation((options, callback) => {
        const location = {
          coords: {
            latitude: 123.456,
            longitude: -98.765,
          },
        };

        callback(location as Location.LocationObject);

        const unsub: Location.LocationSubscription = {
          remove() {},
        };

        return Promise.resolve<Location.LocationSubscription>(unsub);
      });

  const renderGpsCoordinatesInput = () =>
    render(
      <NativeBaseProviderForTesting>
        <GpsCoordinatesInput
          coordinates={{ latitude: undefined, longitude: undefined }}
          usingGps
          onSubmitEditing={onSubmitEditingMock}
          setCoordinates={setCoordinatesMock}
        />
      </NativeBaseProviderForTesting>
    );

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("renders", async () => {
    watchPositionAsyncSuccessMock();
    requestForegroundPermissionsAsyncSuccessMock();

    const { getByDisplayValue, toJSON } = renderGpsCoordinatesInput();

    await waitFor(() => getByDisplayValue("123.456, -98.765"));
    expect(toJSON()).toMatchSnapshot();
  });

  test("permission denied", async () => {
    requestForegroundPermissionsAsyncFailureMock();

    const { getByPlaceholderText } = renderGpsCoordinatesInput();

    await waitFor(() => getByPlaceholderText(Placeholders.GPS.NoPermissions));
  });

  test("failure to watchPositionAsync", async () => {
    jest
      .spyOn(Location, "watchPositionAsync")
      .mockImplementation(() => Promise.reject(Error("mock failure")));

    requestForegroundPermissionsAsyncSuccessMock();

    const { getByPlaceholderText } = renderGpsCoordinatesInput();

    await waitFor(() => getByPlaceholderText(Placeholders.GPS.NoLocation));
  });

  test("extra parenthesis and spaces are clipped", async () => {
    requestForegroundPermissionsAsyncFailureMock();

    const { getByPlaceholderText } = render(
      <NativeBaseProviderForTesting>
        <GpsCoordinatesInput
          coordinates={{ latitude: undefined, longitude: undefined }}
          usingGps
          onSubmitEditing={onSubmitEditingMock}
          setCoordinates={setCoordinatesMock}
        />
      </NativeBaseProviderForTesting>
    );

    await waitFor(() => getByPlaceholderText(Placeholders.GPS.NoPermissions));

    fireEvent.changeText(
      getByPlaceholderText(Placeholders.GPS.NoPermissions),
      "(56.789  , 454.76)   "
    );

    expect(setCoordinatesMock).toBeCalledWith({
      latitude: 56.789,
      longitude: 454.76,
    });
  });

  test("done editing", async () => {
    requestForegroundPermissionsAsyncFailureMock();

    const { getByPlaceholderText } = render(
      <NativeBaseProviderForTesting>
        <GpsCoordinatesInput
          coordinates={{ latitude: undefined, longitude: undefined }}
          usingGps
          onSubmitEditing={onSubmitEditingMock}
          setCoordinates={setCoordinatesMock}
        />
      </NativeBaseProviderForTesting>
    );

    await waitFor(() => getByPlaceholderText(Placeholders.GPS.NoPermissions));
    fireEvent(
      getByPlaceholderText(Placeholders.GPS.NoPermissions),
      "onSubmitEditing"
    );
    expect(onSubmitEditingMock).toHaveBeenCalled();
  });

  test("edit mode with existing coordinates", async () => {
    const foregroundPermissionMock =
      requestForegroundPermissionsAsyncSuccessMock();
    const coordinates = {
      latitude: 33.77112,
      longitude: -133.7331,
    };

    const { getByDisplayValue } = render(
      <NativeBaseProviderForTesting>
        <GpsCoordinatesInput
          coordinates={coordinates}
          usingGps={false}
          onSubmitEditing={onSubmitEditingMock}
          setCoordinates={setCoordinatesMock}
        />
      </NativeBaseProviderForTesting>
    );

    await waitFor(() =>
      getByDisplayValue(`${coordinates.latitude}, ${coordinates.longitude}`)
    );

    expect(foregroundPermissionMock).not.toBeCalled();
  });
});
