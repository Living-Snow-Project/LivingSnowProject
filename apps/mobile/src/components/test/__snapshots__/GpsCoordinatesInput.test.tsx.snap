// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GpsCoordinatesInput tests renders 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    collapsable={false}
    dataSet={Object {}}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={Object {}}
    testID="gps-manual-pressable"
  >
    <View
      dataSet={Object {}}
      pointerEvents="none"
      style={Object {}}
    >
      <View
        dataSet={Object {}}
        isInvalid={false}
        isRequired={true}
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <View
          dataSet={Object {}}
          feedbackId="field-4-feedback"
          hasFeedbackText={true}
          hasHelpText={false}
          helpTextId="field-4-helptext"
          isDisabled={false}
          isInvalid={false}
          isReadOnly={false}
          isRequired={true}
          labelId="field-4-label"
          nativeID="field-4-label"
          setHasFeedbackText={[Function]}
          setHasHelpText={[Function]}
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "flex-start",
              "marginBottom": 4,
              "marginTop": 4,
            }
          }
        >
          <Text
            dataSet={Object {}}
            style={
              Object {
                "backgroundColor": undefined,
                "color": "#737373",
                "fontFamily": undefined,
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "500",
                "letterSpacing": 0,
                "lineHeight": 21,
                "textDecorationLine": undefined,
              }
            }
          >
            GPS Coordinates (latitude, longitude)
          </Text>
          <Text
            dataSet={Object {}}
            style={
              Object {
                "backgroundColor": undefined,
                "color": "#dc2626",
                "fontFamily": undefined,
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "400",
                "letterSpacing": 0,
                "lineHeight": 21,
                "textDecorationLine": undefined,
              }
            }
          >
            *
          </Text>
        </View>
        <View
          dataSet={Object {}}
          isFocused={false}
          style={
            Object {
              "alignItems": "center",
              "borderColor": "#d4d4d4",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexDirection": "row",
              "overflow": "hidden",
            }
          }
        >
          <TextInput
            accessibilityDescribedBy="field-4-feedback"
            accessibilityRequired={true}
            accessible={true}
            blurOnSubmit={false}
            dataSet={Object {}}
            disabled={false}
            editable={true}
            isRequired={true}
            maxLength={30}
            nativeID="field-4-input"
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            onSubmitEditing={[MockFunction]}
            placeholder="Looking for GPS signal"
            placeholderTextColor="#a3a3a3"
            readOnly={false}
            required={true}
            returnKeyType="done"
            secureTextEntry={false}
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "#171717",
                "flex": 1,
                "fontFamily": undefined,
                "fontSize": 16,
                "fontStyle": "normal",
                "fontWeight": "400",
                "height": "100%",
                "paddingBottom": 8,
                "paddingLeft": 12,
                "paddingRight": 12,
                "paddingTop": 8,
                "width": "100%",
              }
            }
            value="123.456, -98.765"
          />
        </View>
      </View>
    </View>
  </View>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "display": "none",
        },
      ]
    }
  >
    <View
      collapsable={false}
      pointerEvents="none"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateY": 0,
            },
            Object {
              "translateX": 0,
            },
            Object {
              "scale": 1,
            },
            Object {
              "scaleX": 1,
            },
            Object {
              "scaleY": 1,
            },
            Object {
              "rotate": "0deg",
            },
          ],
        }
      }
    >
      <View
        accessible={false}
        collapsable={false}
        dataSet={Object {}}
        focusable={true}
        importantForAccessibility="no"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(0, 0, 0)",
            "bottom": 0,
            "left": 0,
            "opacity": 0.3,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
    </View>
    <View
      collapsable={false}
      pointerEvents="none"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateY": 0,
            },
            Object {
              "translateX": 0,
            },
            Object {
              "scale": 1,
            },
            Object {
              "scaleX": 1,
            },
            Object {
              "scaleY": 1,
            },
            Object {
              "rotate": "0deg",
            },
          ],
        }
      }
    >
      <View
        closeOnOverlayClick={true}
        dataSet={Object {}}
        pointerEvents="box-none"
        style={
          Object {
            "alignItems": "center",
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
        testID="gps-enable-manual-coordinates-modal"
      >
        <View
          accessibilityViewIsModal={true}
          aria-modal="true"
          dataSet={Object {}}
          onAccessibilityEscape={[Function]}
          style={
            Object {
              "backgroundColor": "#fafafa",
              "borderRadius": 8,
              "elevation": 2,
              "maxHeight": 1184,
              "maxWidth": 320,
              "overflow": "hidden",
              "shadowColor": "#000000",
              "shadowOffset": Object {
                "height": 1,
                "width": 0,
              },
              "shadowOpacity": 0.2,
              "shadowRadius": 1.41,
              "width": "65%",
            }
          }
        >
          <View
            accessibilityRole="button"
            accessible={true}
            collapsable={false}
            colorScheme="coolGray"
            dataSet={Object {}}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "transparent",
                "borderRadius": 4,
                "paddingBottom": 8,
                "paddingLeft": 8,
                "paddingRight": 8,
                "paddingTop": 8,
                "position": "absolute",
                "right": 12,
                "top": 12,
                "zIndex": 1,
              }
            }
          >
            <RNSVGSvgView
              accessibilityRole="image"
              align="xMidYMid"
              bbHeight="16"
              bbWidth="16"
              dataSet={Object {}}
              focusable={false}
              meetOrSlice={0}
              minX={0}
              minY={0}
              stroke=""
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "borderWidth": 0,
                  },
                  Object {
                    "color": "#737373",
                    "height": 16,
                    "width": 16,
                  },
                  Object {
                    "flex": 0,
                    "height": 16,
                    "width": 16,
                  },
                ]
              }
              tintColor="#737373"
              vbHeight={24}
              vbWidth={24}
            >
              <RNSVGGroup
                propList={
                  Array [
                    "stroke",
                  ]
                }
                stroke={null}
              >
                <RNSVGGroup>
                  <RNSVGPath
                    d="M12 9.77778L4.22222 2L2 4.22222L9.77778 12L2 19.7778L4.22222 22L12 14.2222L19.7778 22L22 19.7778L14.2222 12L22 4.22222L19.7778 2L12 9.77778Z"
                    fill={
                      Object {
                        "type": 2,
                      }
                    }
                    propList={
                      Array [
                        "fill",
                        "stroke",
                      ]
                    }
                    stroke={null}
                  />
                </RNSVGGroup>
              </RNSVGGroup>
            </RNSVGSvgView>
          </View>
          <View
            dataSet={Object {}}
            style={
              Object {
                "backgroundColor": "#fafafa",
                "borderBottomWidth": 1,
                "borderColor": "#d4d4d4",
                "paddingBottom": 16,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 16,
              }
            }
          >
            <Text
              dataSet={Object {}}
              style={
                Object {
                  "backgroundColor": undefined,
                  "color": "#171717",
                  "fontFamily": undefined,
                  "fontSize": 16,
                  "fontStyle": "normal",
                  "fontWeight": "600",
                  "letterSpacing": 0,
                  "lineHeight": 20,
                  "textDecorationLine": undefined,
                }
              }
            >
              Enter coordinates manually?
            </Text>
          </View>
          <RCTScrollView
            contentContainerStyle={
              Array [
                Object {},
                Object {
                  "dataSet": Object {},
                },
                Object {},
              ]
            }
            dataSet={Object {}}
            style={Object {}}
          >
            <View>
              <View
                dataSet={Object {}}
                style={
                  Object {
                    "paddingBottom": 16,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                    "paddingTop": 16,
                  }
                }
              >
                <Text
                  dataSet={Object {}}
                  style={
                    Object {
                      "backgroundColor": undefined,
                      "color": "#171717",
                      "fontFamily": undefined,
                      "fontSize": 14,
                      "fontStyle": "normal",
                      "fontWeight": "400",
                      "letterSpacing": 0,
                      "lineHeight": 21,
                      "textDecorationLine": undefined,
                    }
                  }
                >
                  Some users enter coordinates manually if they cannot acquire GPS signal or when they return home.

This message can be disabled in Settings.
                </Text>
              </View>
            </View>
          </RCTScrollView>
          <View
            dataSet={Object {}}
            style={
              Object {
                "backgroundColor": "#fafafa",
                "borderColor": "#d4d4d4",
                "borderTopWidth": 1,
                "flexDirection": "row",
                "flexWrap": "wrap",
                "justifyContent": "flex-end",
                "paddingBottom": 16,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 16,
              }
            }
          >
            <View
              dataSet={Object {}}
              style={
                Object {
                  "flexDirection": "row",
                }
              }
            >
              <View
                accessibilityRole="button"
                accessible={true}
                collapsable={false}
                colorScheme="blueGray"
                dataSet={Object {}}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "borderRadius": 4,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "paddingBottom": 10,
                    "paddingLeft": 12,
                    "paddingRight": 12,
                    "paddingTop": 10,
                  }
                }
                testID="modal-no-button"
              >
                <View
                  dataSet={Object {}}
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                    }
                  }
                  test={true}
                >
                  <View
                    dataSet={Object {}}
                    style={Object {}}
                  >
                    <Text
                      dataSet={Object {}}
                      style={
                        Object {
                          "backgroundColor": undefined,
                          "color": "#475569",
                          "fontFamily": undefined,
                          "fontSize": 14,
                          "fontStyle": "normal",
                          "fontWeight": "400",
                          "letterSpacing": 0,
                          "lineHeight": 21,
                          "textDecorationLine": undefined,
                        }
                      }
                    >
                      Cancel
                    </Text>
                  </View>
                </View>
              </View>
              <View
                dataSet={Object {}}
                style={
                  Object {
                    "width": 8,
                  }
                }
              />
              <View
                accessibilityRole="button"
                accessible={true}
                collapsable={false}
                colorScheme="primary"
                dataSet={Object {}}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#0891b2",
                    "borderRadius": 4,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "paddingBottom": 10,
                    "paddingLeft": 12,
                    "paddingRight": 12,
                    "paddingTop": 10,
                  }
                }
                testID="modal-confirm-button"
              >
                <View
                  dataSet={Object {}}
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                    }
                  }
                  test={true}
                >
                  <View
                    dataSet={Object {}}
                    style={Object {}}
                  >
                    <Text
                      dataSet={Object {}}
                      style={
                        Object {
                          "backgroundColor": undefined,
                          "color": "#fafafa",
                          "fontFamily": undefined,
                          "fontSize": 14,
                          "fontStyle": "normal",
                          "fontWeight": "400",
                          "letterSpacing": 0,
                          "lineHeight": 21,
                          "textDecorationLine": undefined,
                        }
                      }
                    >
                      Confirm
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
